### PR TITLE
Speed up SharedBytes.IO

### PR DESCRIPTION
**Pull request by original-brownbear**
This was way overengineered and not sure why I built it the way I originally did. We can just put the IO instances in an array of known size and have them live as long as the `SharedBytes` live. No need to constantly create and destroy them and contend on hot map regions whatsoever.
Also, as of right now there is no point in them having individual ref counts. We could probably do away with these in general but I think there's some merit to the API and if we eventually feel like we need to move the IO to mmap these would be a massive help so I didn't make any bigger changes here.

### CODE REVIEW

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java [hunk 1]
```diff
@@ -215,27 +221,11 @@ protected void closeInternal() {
...
-                        decRef();
-                    }
-                }
-                return newIO;
-            }
-            return io;
-        });
+        var res = ios[sharedBytesPos];
+        incRef();
+        return res;
```
Comment (henningandersen) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java:228]: nit: can we avoid the `res` variable?

```suggestion
        incRef();
        return ios[sharedBytesPos];
```

**Review by henningandersen — APPROVED** [review]
LGTM.

Can we add a small test of the release part? Clearly works, but would be good to have a test anyway. Just something like:

1. create shared bytes.
2. getFileChannel
3. Close IO from getFileChannel.
4. close shared bytes
5. Check file is deleted (and not sooner).

(perhaps a variant where 3 and 4 are swapped too).

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java (hunk 1 above)
Comment (original-brownbear) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java:228]: I'd rather not, if I ever have to debug a leak in this thing, I'll waste time hunting down whether or not I ever got an out of bounds exception here :)

Comment (henningandersen) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java:228]: ah, makes sense, thanks.

**Comment by original-brownbear** [comment]
Thanks Henning! + added tests as requested :)

**Pull request merged**